src/conduits/RwaSwapInputConduit2.sol

```diff
@@ -0,0 +1,329 @@
+        // Give unlimited approval to PSM gemjoin
+        GemAbstract(_gem).approve(address(psm.gemJoin()), type(uint256).max);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    /*//////////////////////////////////
+               Authorization
+    //////////////////////////////////*/
+
+    /**
+     * @notice Grants `usr` admin access to this contract.
+     * @param usr The user address.
+     */
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    /**
+     * @notice Revokes `usr` admin access from this contract.
+     * @param usr The user address.
+     */
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    /**
+     * @notice Grants `usr` push access to this contract.
+     * @param usr The user address.
+     */
+    function mate(address usr) external auth {
+        may[usr] = 1;
+        emit Mate(usr);
+    }
+
+    /**
+     * @notice Revokes `usr` push access from this contract.
+     * @param usr The user address.
+     */
+    function hate(address usr) external auth {
+        may[usr] = 0;
+        emit Hate(usr);
+    }
+
+    /*//////////////////////////////////
+               Administration
+    //////////////////////////////////*/
+
+    /**
+     * @notice Updates a contract parameter.
+     * @param what The changed parameter name. `"to", "quitTo", "psm"`
+     * @param data The new value of the parameter.
+     */
+    function file(bytes32 what, address data) external auth {
+        if (what == "quitTo") {
+            quitTo = data;
+        } else if (what == "to") {
+            to = data;
+        } else if (what == "psm") {
+            require(PsmAbstract(data).dai() == address(dai), "RwaSwapInputConduit2/wrong-dai-for-psm");
+            require(
+                GemJoinAbstract(PsmAbstract(data).gemJoin()).gem() == address(gem),
+                "RwaSwapInputConduit2/wrong-gem-for-psm"
+            );
+
+            // Revoke approval for the old PSM gemjoin
+            gem.approve(address(psm.gemJoin()), 0);
+            // Give unlimited approval to the new PSM gemjoin
+            gem.approve(address(PsmAbstract(data).gemJoin()), type(uint256).max);
+
+            psm = PsmAbstract(data);
+        } else {
+            revert("RwaSwapInputConduit2/unrecognised-param");
+        }
+
+        emit File(what, data);
+    }
+
+    /*//////////////////////////////////
+               Operations
+    //////////////////////////////////*/
+
+    /**
+     * @notice Swaps the GEM balance of this contract into DAI through the PSM and push it into the recipient address.
+     */
+    function push() external {
+        _doPush(gem.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Swaps the specified amount of GEM into DAI through the PSM and push it into the recipient address.
+     * @param amt Gem amount.
+     */
+    function push(uint256 amt) external {
```

Comment:
Are the parametrized versions of `push` and `quit` really needed? I'm struggling to come up with a use case.

Reply:
The base use case is `push`ing the entire balance of the contract into the RwaUrn. That's the `push()` function without any arguments.

However, we might have a case where the PSM is nearly full. If the balance sitting in the contract is greater than what can be put into the PSM, the transaction will fail. We added this as a fallback mechanism in case only a given amount of tokens can be swapped into Dai for repayment.

From there, it could go 2 ways:
1. The PSM gains some room after a while and the `push()` function can be called again
2. Governance can withdraw `gem` from the contract through `yank`.

Reply:
`quit` is another escape hatch in case the wrong amount of `gem` is sent to the contract.

1. If the PSM is full, the operator might need to claim tokens back through `quit()`
2. If someone mistyped the number, they can claim the difference back through `quit(uint)`

The idea is not require governance intervention in such cases.

Reply:
Thanks, makes sense. Ofc such situations could be remedied in other ways but the added complexity isn't too much so might as well keep the flexibility.